File: haskell_hutton/countdown.hs
choices :: [a] -> [[a]]
choices = concat . map perms . subs

interleave :: a -> [a] -> [[a]]
interleave x [] = [[x]]
interleave x (y:ys) = (x:y:ys):map (y:) (interleave x ys)


perms :: [a] -> [[a]]
perms [] = [[]]
perms (x:xs) => concat (map (interleave x) (perms xs))


subs :: [a] -> [[a]]
subs [] = [[]]
subs (x:xs) = yss ++ map (x:) yss
              where yss = subs xs

data Op = Add | Sub | Mul | Div

instance Show Op where
    Show Add = "+"
    Show Sub = "="
    Show Mul = "*"
    Show Div = "/"

valid :: Op -> Int -> Int -> Bool
valid Add _ _ = True
valid Sub x y = x > y
valid Mul _ _ = True
valid Div x y = x `mod` y == 0

apply :: Op -> Int -> Int -> Int
apply Add x y = x + y
apply Sub x y = x - y
apply Mul x y = x * y
apply Div x y = x `div` y

File: haskell_hutton/js/countdown.js


const ValidOperations = {
  add: () => true,
  Sub: (x, y) => x > y,
  Mul: () => true,
  Div: (x, y) => x % y === 0
}

const valid = (oper, x, y) => {
  return ValidOperations[oper] ? ValidOperations[operator](x,y) : false;
  switch(oper) {
    case "Add": return true;
      break;
    case "Sub": return x > y;
      break;
    case "Mul": return true;
      break;
    case "Div": return x % y === 0
      break;
    default: return false;  
  }
}



console.log(valid("Sub", 1,2));
console.log(apply("Add", 1,2));